pkg/controller/core.oam.dev/v1alpha2/application/apply.go

```diff
@@ -287,122 +204,16 @@ func (h *appHandler) statusAggregate(appFile *appfile.Appfile) ([]common.Applica
 	return appStatus, healthy, nil
 }
 
-// createOrUpdateComponent creates a component if not exist and update if exists.
-// it returns the corresponding component revisionName and if a new component revision is created
-func (h *appHandler) createOrUpdateComponent(ctx context.Context, comp *v1alpha2.Component) (string, error) {
-	curComp := v1alpha2.Component{}
-	var preRevisionName, curRevisionName string
-	compName := comp.GetName()
-	compNameSpace := comp.GetNamespace()
-	compKey := ctypes.NamespacedName{Name: compName, Namespace: compNameSpace}
-
-	err := h.r.Get(ctx, compKey, &curComp)
-	if err != nil {
-		if !apierrors.IsNotFound(err) {
-			return "", err
-		}
-		if err = h.r.Create(ctx, comp); err != nil {
-			return "", err
-		}
-		klog.InfoS("Created a new component", "component", klog.KObj(comp))
-	} else {
-		// remember the revision if there is a previous component
-		if curComp.Status.LatestRevision != nil {
-			preRevisionName = curComp.Status.LatestRevision.Name
-		}
-		comp.ResourceVersion = curComp.ResourceVersion
-		if err := h.r.Update(ctx, comp); err != nil {
-			return "", err
-		}
-		klog.InfoS("Updated a component", "component", klog.KObj(comp))
-	}
-	// remove the object from the raw extension before we can compare with the existing componentRevision whose
-	// object is persisted as Raw data after going through api server
-	updatedComp := comp.DeepCopy()
-	updatedComp.Spec.Workload.Object = nil
-	if updatedComp.Spec.Helm != nil {
-		updatedComp.Spec.Helm.Release.Object = nil
-		updatedComp.Spec.Helm.Repository.Object = nil
-	}
-	if len(preRevisionName) != 0 {
-		needNewRevision, err := utils.CompareWithRevision(ctx, h.r, compName, compNameSpace, preRevisionName, &updatedComp.Spec)
-		if err != nil {
-			return "", errors.Wrap(err, fmt.Sprintf("compare with existing controllerRevision %s failed",
-				preRevisionName))
-		}
-		if !needNewRevision {
-			klog.InfoS("No need to wait for a new component revision", "component", klog.KObj(updatedComp),
-				"revision", preRevisionName)
-			return preRevisionName, nil
-		}
-	}
-	klog.InfoS("Wait for a new component revision", "component name", compName,
-		"previous revision", preRevisionName)
-	// get the new component revision that contains the component with retry
-	checkForRevision := func() (bool, error) {
-		if err := h.r.Get(ctx, compKey, &curComp); err != nil {
-			// retry no matter what
-			// nolint:nilerr
-			return false, nil
-		}
-		if curComp.Status.LatestRevision == nil || curComp.Status.LatestRevision.Name == preRevisionName {
-			return false, nil
-		}
-		needNewRevision, err := utils.CompareWithRevision(ctx, h.r, compName,
```

Comment:
I think the logic is still valuable, we may need component level revision in the future. This can be refactored to  compare and create revision for component resources

Reply:
replace it with new component revision mechanism. We will no longer emit v1alpha2.Component into K8s and let ComponentHandler to generate component revision.  Instead, we create `ControllerRevision` in app controller while handling component revision. And latest component revision is recorded in latest app revision's AppConfig's AppConfigComponent.